File: faas/internal/tododraft/tododraft_test.go
package tododraft

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"testing"

	"github.com/mdk/digitaltwin2026/faas/internal/record"
	"github.com/mdk/digitaltwin2026/faas/internal/tags"
)

type deformFixture struct {
	InputRecord    record.Record  `json:"inputRecord"`
	TodoRecordJSON TodoRecordJSON `json:"todoRecordJson"`
}

func repoRoot(t *testing.T) string {
	t.Helper()
	_, file, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatal("runtime.Caller failed")
	}
	return filepath.Clean(filepath.Join(filepath.Dir(file), "..", "..", ".."))
}

func TestToTodoRecordJSONSharedFixture(t *testing.T) {
	b, err := os.ReadFile(filepath.Join(repoRoot(t), "testdata", "todo-record-deform.json"))
	if err != nil {
		t.Fatal(err)
	}
	var fx deformFixture
	if err := json.Unmarshal(b, &fx); err != nil {
		t.Fatal(err)
	}
	got := ToTodoRecordJSON(fx.InputRecord)
	gotJSON, err := json.Marshal(got)
	if err != nil {
		t.Fatal(err)
	}
	wantJSON, err := json.Marshal(fx.TodoRecordJSON)
	if err != nil {
		t.Fatal(err)
	}
	if string(gotJSON) != string(wantJSON) {
		t.Fatalf("got=%s want=%s", gotJSON, wantJSON)
	}
}

func TestToTodoRecordJSONPreservesOffset(t *testing.T) {
	rec := record.Record{
		ID:               "01900000-0000-7000-8000-000000000003",
		HappenedAt:       "2026-08-02T10:00:00.000+08:00",
		RawContent:       strPtr("Buy milk"),
		Tags:             []string{"todo:in_progress", "errand"},
		ObjectiveContext: "x",
	}
	got := ToTodoRecordJSON(rec)
	if got.CreatedAt != "2026-08-02T10:00:00.000+08:00" {
		t.Fatalf("created_at=%q", got.CreatedAt)
	}
}

func strPtr(s string) *string { return &s }

func TestShouldDeformTodoRecordTagsSharedFixture(t *testing.T) {
	b, err := os.ReadFile(filepath.Join(repoRoot(t), "testdata", "todo-query-deform-cases.json"))
	if err != nil {
		t.Fatal(err)
	}
	var fx struct {
		Cases []struct {
			Name   string   `json:"name"`
			Tags   []string `json:"tags"`
			Deform bool     `json:"deform"`
		} `json:"cases"`
	}
	if err := json.Unmarshal(b, &fx); err != nil {
		t.Fatal(err)
	}
	for _, c := range fx.Cases {
		got := ShouldDeformTodoRecordTags(c.Tags)
		if got != c.Deform {
			t.Fatalf("%s: got=%v want=%v tags=%v", c.Name, got, c.Deform, c.Tags)
		}
	}
}

func TestParseTodo(t *testing.T) {
	raw := []byte(`{
		"created_at": "2026-08-02T10:00:00+08:00",
		"content": "Buy milk",
		"objective_context": "weekend grocery list",
		"ai_analysis": "need it for breakfast",
		"tags": ["errand"]
	}`)
	got, err := ParseTodo(raw)
	if err != nil {
		t.Fatal(err)
	}
	if got.RawContent != "Buy milk" {
		t.Fatalf("content=%q", got.RawContent)
	}
	if len(got.Tags) != 2 || got.Tags[0] != TodoTagInProgress || got.Tags[1] != "errand" {
		t.Fatalf("tags=%v", got.Tags)
	}
	if got.HappenedAtRaw != "2026-08-02T10:00:00+08:00" {
		t.Fatalf("happenedAtRaw=%q", got.HappenedAtRaw)
	}
}

func TestParseTodoOmitsTags(t *testing.T) {
	raw := []byte(`{
		"created_at": "2026-08-02T10:00:00+08:00",
		"content": "Buy milk",
		"objective_context": "x"
	}`)
	got, err := ParseTodo(raw)
	if err != nil {
		t.Fatal(err)
	}
	if len(got.Tags) != 1 || got.Tags[0] != TodoTagInProgress {
		t.Fatalf("tags=%v", got.Tags)
	}
}

func TestParseTodoRejects(t *testing.T) {
	cases := []struct {
		raw  string
		want string
	}{
		{
			`{"content":"x","objective_context":"y"}`,
			"missing required field: created_at",
		},
		{
			`{"created_at":"2026-08-02T10:00:00+08:00","objective_context":"y"}`,
			"missing required field: content",
		},
		{
			`{"created_at":"2026-08-02T10:00:00+08:00","content":"x"}`,
			"missing required field: objective_context",
		},
		{
			`{"created_at":"2026-08-02T10:00:00","content":"x","objective_context":"y"}`,
			"created_at must be ISO 8601 with timezone (Z or ±HH:MM)",
		},
		{
			`{"created_at":"2026-08-02T10:00:00+08:00","content":"x","objective_context":"y","tags":["todo:in_progress"]}`,
			tags.ReservedTagError("todo:in_progress"),
		},
		{
			`{"created_at":"2026-08-02T10:00:00+08:00","happened_at":"2026-08-02T10:00:00+08:00","content":"x","objective_context":"y"}`,
			"Unknown JSON key: happened_at",
		},
		{
			`{"created_at":"2026-08-02T10:00:00+08:00","content":"x","objective_context":"y","raw_content":"x"}`,
			"Unknown JSON key: raw_content",
		},
	}
	for _, tc := range cases {
		_, err := ParseTodo([]byte(tc.raw))
		if err == nil || err.Error() != tc.want {
			t.Fatalf("raw=%s err=%v want=%q", tc.raw, err, tc.want)
		}
	}
}

func TestIsStrictTodoRecordTags(t *testing.T) {
	if !IsStrictTodoRecordTags([]string{TodoTagInProgress, "errand"}) {
		t.Fatal("expected strict todo")
	}
	if IsStrictTodoRecordTags([]string{TodoTagTransition}) {
		t.Fatal("audit must not be strict todo")
	}
	if IsStrictTodoRecordTags([]string{TodoTagInProgress, TodoTagCompleted}) {
		t.Fatal("two states must fail")
	}
	if IsStrictTodoRecordTags([]string{"errand"}) {
		t.Fatal("no state must fail")
	}
}

func TestTodoAuditNotifyTextSharedFixture(t *testing.T) {
	b, err := os.ReadFile(filepath.Join(repoRoot(t), "testdata", "todo-transition-audit.json"))
	if err != nil {
		t.Fatal(err)
	}
	var fx struct {
		Cases []struct {
			Target           string `json:"target"`
			TodoID           string `json:"todoId"`
			TodoHappenedAt   string `json:"todoHappenedAt"`
			TodoRawContent   string `json:"todoRawContent"`
			ObjectiveContext string `json:"objective_context"`
			NotifyText       string `json:"notify_text"`
		} `json:"cases"`
	}
	if err := json.Unmarshal(b, &fx); err != nil {
		t.Fatal(err)
	}
	for _, c := range fx.Cases {
		got := TodoAuditNotifyText(c.Target, c.TodoID, c.TodoHappenedAt, c.TodoRawContent)
		if got != c.NotifyText {
			t.Fatalf("target=%s notify got=%q want=%q", c.Target, got, c.NotifyText)
		}
		gotObj := AuditObjectiveContext(c.Target, c.TodoID, c.TodoHappenedAt)
		if gotObj != c.ObjectiveContext {
			t.Fatalf("target=%s objective got=%q want=%q", c.Target, gotObj, c.ObjectiveContext)
		}
	}
}

func TestReplaceTodoStateInTags(t *testing.T) {
	got := ReplaceTodoStateInTags([]string{TodoTagInProgress, "errand"}, TodoStateCompleted)
	if len(got) != 2 || got[0] != TodoTagCompleted || got[1] != "errand" {
		t.Fatalf("got=%v", got)
	}
	if !IsTodoAuditRecordTags([]string{TodoTagTransition}) {
		t.Fatal("expected audit")
	}
	if TodoStateFromTags([]string{TodoTagInProgress, "errand"}) != TodoStateInProgress {
		t.Fatal("expected in_progress")
	}
}

func TestParseTodoTransition(t *testing.T) {
	raw := []byte(`{
		"id": "01900000-0000-7000-8000-000000000003",
		"target": "completed",
		"happened_at": "2026-08-02T12:00:00+08:00"
	}`)
	got, err := ParseTodoTransition(raw)
	if err != nil {
		t.Fatal(err)
	}
	if got.ID != "01900000-0000-7000-8000-000000000003" || got.Target != TodoStateCompleted {
		t.Fatalf("got=%+v", got)
	}
	if got.HappenedAtRaw != "2026-08-02T12:00:00+08:00" {
		t.Fatalf("happenedAtRaw=%q", got.HappenedAtRaw)
	}
}

func TestParseTodoTransitionRejects(t *testing.T) {
	cases := []struct {
		raw  string
		want string
	}{
		{
			`{"target":"completed","happened_at":"2026-08-02T12:00:00+08:00"}`,
			"missing required field: id",
		},
		{
			`{"id":"01900000-0000-7000-8000-000000000003","happened_at":"2026-08-02T12:00:00+08:00"}`,
			"missing required field: target",
		},
		{
			`{"id":"01900000-0000-7000-8000-000000000003","target":"completed"}`,
			"missing required field: happened_at",
		},
		{
			`{"id":"01900000-0000-7000-8000-000000000003","target":"done","happened_at":"2026-08-02T12:00:00+08:00"}`,
			ErrInvalidTarget.Error(),
		},
		{
			`{"id":"01900000-0000-7000-8000-000000000003","target":"completed","happened_at":"2026-08-02T12:00:00+08:00","created_at":"x"}`,
			"Unknown JSON key: created_at",
		},
	}
	for _, tc := range cases {
		_, err := ParseTodoTransition([]byte(tc.raw))
		if err == nil || err.Error() != tc.want {
			t.Fatalf("raw=%s err=%v want=%q", tc.raw, err, tc.want)
		}
	}
}

func TestParseTodoRejectsDuplicateTags(t *testing.T) {
	raw := []byte(`{
		"created_at": "2026-08-02T10:00:00+08:00",
		"content": "Buy milk",
		"objective_context": "x",
		"tags": ["errand", "errand"]
	}`)
	_, err := ParseTodo(raw)
	if err == nil || err.Error() != `duplicate tag "errand"` {
		t.Fatalf("err: %v", err)
	}
}

File: src/lib/tododraft.test.ts
import { readFileSync } from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'
import { describe, expect, it } from 'vitest'
import { reservedTagError } from './tags'
import {
  auditObjectiveContext,
  todoAuditNotifyText,
  ERR_INVALID_TARGET,
  isTodoAuditRecordTags,
  parseTodo,
  parseTodoTransition,
  replaceTodoStateInTags,
  shouldDeformTodoRecordTags,
  toTodoRecordJson,
  TODO_TAG_IN_PROGRESS,
  TODO_TAG_TRANSITION,
  todoStateFromTags,
  type TodoRecordJson,
  type TodoState,
} from './tododraft'
import type { Record } from './record'
import { toQueryRecordJson } from './query'

const deformFixture = JSON.parse(
  readFileSync(
    path.join(
      path.dirname(fileURLToPath(import.meta.url)),
      '../../testdata/todo-record-deform.json',
    ),
    'utf8',
  ),
) as {
  inputRecord: Record
  todoRecordJson: TodoRecordJson
}

describe('toTodoRecordJson shared fixture', () => {
  it('maps happenedAt/rawContent to created_at/content', () => {
    expect(toTodoRecordJson(deformFixture.inputRecord)).toEqual(
      deformFixture.todoRecordJson,
    )
  })

  it('preserves offset on created_at (deform is key-only, §6.1)', () => {
    const withOffset = {
      ...deformFixture.inputRecord,
      happened_at: '2026-08-02T10:00:00.000+08:00',
    }
    expect(toTodoRecordJson(withOffset).created_at).toBe(
      '2026-08-02T10:00:00.000+08:00',
    )
  })
})

const queryDeformFixture = JSON.parse(
  readFileSync(
    path.join(
      path.dirname(fileURLToPath(import.meta.url)),
      '../../testdata/todo-query-deform-cases.json',
    ),
    'utf8',
  ),
) as {
  cases: Array<{ name: string; tags: string[]; deform: boolean }>
}

describe('shouldDeformTodoRecordTags shared fixture', () => {
  it('matches query-side wide rule for all cases', () => {
    for (const c of queryDeformFixture.cases) {
      expect(shouldDeformTodoRecordTags(c.tags), c.name).toBe(c.deform)
    }
  })
})

describe('toQueryRecordJson', () => {
  it('deforms todo rows and keeps audit / plain as Record', () => {
    const todo = toQueryRecordJson(deformFixture.inputRecord)
    expect(todo).toEqual(deformFixture.todoRecordJson)
    expect('happened_at' in todo).toBe(false)
    expect('raw_content' in todo).toBe(false)

    const audit: Record = {
      ...deformFixture.inputRecord,
      raw_content: 'Buy milk',
      tags: [TODO_TAG_TRANSITION],
      objective_context:
        'Complete a to-do 01900000-0000-7000-8000-000000000003 created at 2026-08-02T02:00:00.000Z',
      ai_analysis: null,
    }
    const auditJson = toQueryRecordJson(audit)
    expect(auditJson).toEqual(audit)
    expect('created_at' in auditJson).toBe(false)
    expect('content' in auditJson).toBe(false)

    const dirty: Record = {
      ...deformFixture.inputRecord,
      tags: ['todo:completed', TODO_TAG_TRANSITION],
    }
    const dirtyJson = toQueryRecordJson(dirty)
    expect(dirtyJson).toMatchObject({
      created_at: dirty.happened_at,
      content: dirty.raw_content,
    })
    expect('happened_at' in dirtyJson).toBe(false)
  })
})

describe('parseTodo', () => {
  const base = {
    created_at: '2026-08-02T10:00:00+08:00',
    content: 'Buy milk',
    objective_context: 'weekend grocery list',
  }

  it('prepends todo:in_progress and maps aliases', () => {
    const parsed = parseTodo({
      ...base,
      ai_analysis: 'need it for breakfast',
      tags: ['errand'],
    })
    expect('error' in parsed).toBe(false)
    if ('error' in parsed) return
    expect(parsed.rawContent).toBe('Buy milk')
    expect(parsed.tags).toEqual([TODO_TAG_IN_PROGRESS, 'errand'])
    expect(parsed.objectiveContext).toBe('weekend grocery list')
    expect(parsed.aiAnalysis).toBe('need it for breakfast')
    expect(parsed.happenedAt.toISOString()).toBe('2026-08-02T02:00:00.000Z')
  })

  it('allows omitted or empty tags (only todo:in_progress)', () => {
    const a = parseTodo(base)
    expect('error' in a).toBe(false)
    if ('error' in a) return
    expect(a.tags).toEqual([TODO_TAG_IN_PROGRESS])

    const b = parseTodo({ ...base, tags: [] })
    expect('error' in b).toBe(false)
    if ('error' in b) return
    expect(b.tags).toEqual([TODO_TAG_IN_PROGRESS])
  })

  it('rejects missing created_at / content / objective_context', () => {
    expect(parseTodo({ content: 'x', objective_context: 'y' })).toEqual({
      error: 'Missing required field: created_at',
    })
    expect(parseTodo({ created_at: base.created_at, objective_context: 'y' })).toEqual({
      error: 'Missing required field: content',
    })
    expect(parseTodo({ created_at: base.created_at, content: 'x' })).toEqual({
      error: 'Missing required field: objective_context',
    })
    expect(
      parseTodo({ ...base, content: '' }),
    ).toEqual({ error: 'Missing required field: content' })
  })

  it('rejects created_at without timezone using created_at wording', () => {
    expect(
      parseTodo({ ...base, created_at: '2026-08-02T10:00:00' }),
    ).toEqual({
      error: 'created_at must be ISO 8601 with timezone (Z or ±HH:MM)',
    })
  })

  it('rejects reserved client tags', () => {
    expect(parseTodo({ ...base, tags: ['todo:in_progress'] })).toEqual({
      error: reservedTagError('todo:in_progress'),
    })
    expect(parseTodo({ ...base, tags: ['todo'] })).toEqual({
      error: reservedTagError('todo'),
    })
  })

  it('rejects duplicate client tags', () => {
    expect(parseTodo({ ...base, tags: ['errand', 'errand'] })).toEqual({
      error: 'Duplicate tag "errand"',
    })
  })

  it('rejects happened_at / raw_content as unknown keys', () => {
    expect(
      parseTodo({ ...base, happened_at: base.created_at } as typeof base & {
        happened_at: string
      }),
    ).toEqual({ error: 'Unknown JSON key: happened_at' })
    expect(
      parseTodo({ ...base, raw_content: 'x' } as typeof base & {
        raw_content: string
      }),
    ).toEqual({ error: 'Unknown JSON key: raw_content' })
  })
})

const auditFixture = JSON.parse(
  readFileSync(
    path.join(
      path.dirname(fileURLToPath(import.meta.url)),
      '../../testdata/todo-transition-audit.json',
    ),
    'utf8',
  ),
) as {
  cases: Array<{
    target: TodoState
    todoId: string
    todoHappenedAt: string
    todoRawContent: string
    objective_context: string
    notify_text: string
  }>
}

describe('todoAuditNotifyText shared fixture', () => {
  it('matches §3.1 notify templates for all targets', () => {
    for (const c of auditFixture.cases) {
      expect(
        todoAuditNotifyText(
          c.target,
          c.todoId,
          c.todoHappenedAt,
          c.todoRawContent,
        ),
      ).toBe(c.notify_text)
    }
  })

  it('builds audit objective_context with verb, todo id and time', () => {
    for (const c of auditFixture.cases) {
      expect(
        auditObjectiveContext(c.target, c.todoId, c.todoHappenedAt),
      ).toBe(c.objective_context)
    }
  })
})

describe('todo transition tag helpers', () => {
  it('detects audit vs strict todo and replaces state', () => {
    expect(isTodoAuditRecordTags([TODO_TAG_TRANSITION])).toBe(true)
    expect(isTodoAuditRecordTags([TODO_TAG_IN_PROGRESS])).toBe(false)
    expect(todoStateFromTags([TODO_TAG_IN_PROGRESS, 'errand'])).toBe(
      'in_progress',
    )
    expect(
      replaceTodoStateInTags([TODO_TAG_IN_PROGRESS, 'errand'], 'completed'),
    ).toEqual(['todo:completed', 'errand'])
  })
})

describe('parseTodoTransition', () => {
  const base = {
    id: '01900000-0000-7000-8000-000000000003',
    target: 'completed',
    happened_at: '2026-08-02T12:00:00+08:00',
  }

  it('accepts valid transition body', () => {
    const parsed = parseTodoTransition(base)
    expect('error' in parsed).toBe(false)
    if ('error' in parsed) return
    expect(parsed.id).toBe(base.id)
    expect(parsed.target).toBe('completed')
    expect(parsed.happenedAt.toISOString()).toBe('2026-08-02T04:00:00.000Z')
  })

  it('rejects missing fields / invalid target / unknown keys', () => {
    expect(parseTodoTransition({ ...base, id: undefined })).toEqual({
      error: 'Missing required field: id',
    })
    expect(parseTodoTransition({ ...base, target: undefined })).toEqual({
      error: 'Missing required field: target',
    })
    expect(parseTodoTransition({ ...base, happened_at: undefined })).toEqual({
      error: 'Missing required field: happened_at',
    })
    expect(parseTodoTransition({ ...base, target: 'done' })).toEqual({
      error: ERR_INVALID_TARGET,
    })
    expect(
      parseTodoTransition({
        ...base,
        created_at: base.happened_at,
      } as typeof base & { created_at: string }),
    ).toEqual({ error: 'Unknown JSON key: created_at' })
  })
})
